analyzer/codechecker_analyzer/analyzers/clangsa/ctu_autodetection.py

```diff
@@ -0,0 +1,134 @@
+#   License. See LICENSE.TXT for details.
+# -------------------------------------------------------------------------
+"""
+Clang Static Analyzer related functions.
+"""
+
+from __future__ import print_function
+from __future__ import division
+from __future__ import absolute_import
+
+import os
+import re
+import subprocess
+
+from codechecker_common.logger import get_logger
+
+LOG = get_logger('analyzer.clangsa')
+
+
+def test_binary_existence(tool_path):
+    """ Test the executable for existence by checking its version. """
+    try:
+        version = subprocess.check_output([tool_path, '-version'])
+    except (subprocess.CalledProcessError, OSError):
+        version = 'ERROR'
+    return version != 'ERROR'
+
+
+class CTUAutodetection(object):
+
+    def __init__(self, analyzer_binary):
+        self.__analyzer_binary = analyzer_binary
+        self.__analyzer_version_info = None
+
+    @property
+    def analyzer_version_info(self):
+        """ Returns the relevant parameters of the analyzer by parsing the
+        output of the analyzer binary when called with version flag. """
+
+        if self.__analyzer_binary is None:
+            return False
+
+        if self.__analyzer_version_info is not None:
+            return self.__analyzer_version_info
+
+        try:
+            analyzer_output = subprocess.check_output(
+                [self.__analyzer_binary, '--version'])
+        except (subprocess.CalledProcessError, OSError):
+            return False
+
+        version_pattern = re.compile(
+            r'clang version (?P<major_version>[0-9]+)'
+            r'\.(?P<minor_version>[0-9]+)\.(?P<patch_version>[0-9]+)\s+'
+            r'\((?P<clang_repository>[^\s]*)\s+(?P<clang_commit>[^\s]*)\)\s+'
+            r'\((?P<llvm_repository>[^\s]*)\s+(?P<llvm_commit>[^\s]*)\)\s+'
+            r'Target: (?P<target>[^\s]*)\s+'
+            r'Thread model: (?P<thread_model>[^\s]*)\s+'
+            r'InstalledDir: (?P<installed_dir>[^\s]*).*')
+
+        match = (version_pattern.search(analyzer_output))
+
+        if not match:
+            return False
+
+        self.__analyzer_version_info = match.groupdict()
+        return self.__analyzer_version_info
+
+    @property
+    def major_version(self):
+        """ Returns the major version of the analyzer, which is used for
+        CTU analysis. """
+
+        if not self.analyzer_version_info:
+            return False
+
+        try:
+            return (int)(self.analyzer_version_info['major_version'])
+        except ValueError:
+            return False
+
+    @property
+    def installed_dir(self):
+        """ Returns the installed directory of the analyzer, which is used for
+        CTU analysis. """
+
+        if not self.analyzer_version_info:
+            return False
+
+        return self.analyzer_version_info['installed_dir']
+
+    @property
+    def mapping_tool_path(self):
+        """ Returns the path of the mapping tool, which is assumed to be
```

Comment:
Is it always available next to clang? If you run a standard clang install on ubuntu (apt-get install clang) will the extra binaries installed too? (`clang-extdef-mapping`, `clang-func-mapping`)

Reply:
I have extenden the implementation to include clang-extdef-mapping, clang-func-mapping, clang-extdef-mapping-<major_version> and clang-func-mapping-<major_version> alternatives.